# Exercise 1

In [50]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import altair as alt

df = pd.read_csv("https://media.githubusercontent.com/media/nickeubank/MIDS_Data/master/OnlineArrestData1980-2021.csv")
df.head(10)


YEAR GENDER      RACE    AGE_GROUP          COUNTY  VIOLENT  PROPERTY  \
0  1980   Male     Black     Under 18  Alameda County      505      1351   
1  1980   Male     Black     18 to 19  Alameda County      205       465   
2  1980   Male     Black     20 to 29  Alameda County      949      1593   
3  1980   Male     Black     30 to 39  Alameda County      450       755   
4  1980   Male     Black     40 to 69  Alameda County      172       218   
5  1980   Male     Black  70 and over  Alameda County        4         2   
6  1980   Male  Hispanic     Under 18  Alameda County      151       365   
7  1980   Male  Hispanic     18 to 19  Alameda County       83        89   
8  1980   Male  Hispanic     20 to 29  Alameda County      226       243   
9  1980   Male  Hispanic     30 to 39  Alameda County       59        80   

   F_DRUGOFF  F_SEXOFF  F_ALLOTHER  F_TOTAL  M_TOTAL  S_TOTAL  
0        188        26          79     2149     2286      295  
1        183         8          48      909     1333        0  
2        606        27         178     3353     7974        0  
3        241        18         110     1574     4876        0  
4        117        11          66      584     3836        0  
5          2         0           0        8       79        0  
6         40         9          77      642     1303      187  
7         48         0          58      278      963        0  
8        158        13          86      726     3748        0  
9         41         8          33      221     1669        0

# Exercise 2

From the table, each row represent how many people have the relavant data values, for example, under F_drugoff, there are 188 felony males under age 18 in Alameda County. 

# Exericse 3

In [51]:
df1 = df.duplicated(["YEAR", "GENDER", "RACE", "AGE_GROUP", "COUNTY"]).sum()
assert not df1.any()

# Exercise 4

In [52]:
ex4_num_rows = df.groupby(["YEAR", "RACE", "COUNTY"], as_index=False).sum()
ex4_num_rows

/var/folders/k7/mj4xq7n54ls4p8zkcgghbg8c0000gn/T/ipykernel_66535/483026680.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  ex4_num_rows = df.groupby(["YEAR", "RACE", "COUNTY"], as_index=False).sum()


YEAR   RACE               COUNTY  VIOLENT  PROPERTY  F_DRUGOFF  \
0     1980  Black       Alameda County     2594      5138       1489   
1     1980  Black         Butte County       32        33          5   
2     1980  Black     Calaveras County        0         0          0   
3     1980  Black        Colusa County        0         5          0   
4     1980  Black  Contra Costa County      720      1422        368   
...    ...    ...                  ...      ...       ...        ...   
9697  2021  White        Tulare County      524       407        195   
9698  2021  White      Tuolumne County      182        63         35   
9699  2021  White       Ventura County      574       372        308   
9700  2021  White          Yolo County      171       163         51   
9701  2021  White          Yuba County      215       123         68   

      F_SEXOFF  F_ALLOTHER  F_TOTAL  M_TOTAL  S_TOTAL  
0           94         541     9856    24081      555  
1            0           7       77      139        0  
2            0           0        0        7        0  
3            0           0        5       13        0  
4           24         257     2791     5201       40  
...        ...         ...      ...      ...      ...  
9697        10         413     1549     3368       31  
9698        13         289      582     1393        2  
9699        13         672     1939     7509       18  
9700         8         173      566     1051        0  
9701        20         138      564     1160        0  

[9702 rows x 11 columns]

# Exercise 7

In [56]:
ex4_num_rows[(ex4_num_rows["COUNTY"] == "Alameda County") & (ex4_num_rows["YEAR"] == 1980)]


YEAR      RACE          COUNTY  VIOLENT  PROPERTY  F_DRUGOFF  F_SEXOFF  \
0    1980     Black  Alameda County     2594      5138       1489        94   
55   1980  Hispanic  Alameda County      592       903        361        34   
112  1980     Other  Alameda County      139       233         69        10   
170  1980     White  Alameda County     1179      3103       1650       119   

     F_ALLOTHER  F_TOTAL  M_TOTAL  S_TOTAL  total_violent_by_race  \
0           541     9856    24081      555                   4504   
55          279     2169    10120      333                   4504   
112          64      515     2686       92                   4504   
170         767     6818    34455     1358                   4504   

     total_drug_by_race  
0                  3569  
55                 3569  
112                3569  
170                3569

# Exercise 5

In [54]:
ex5_collapsed_vars = ["AGE_GROUP", "GENDER"]

# Exercise 6

In [55]:
violent_year = ex4_num_rows.groupby(['COUNTY', 'YEAR'])['VIOLENT'].transform('sum')
drug_year = ex4_num_rows.groupby(['COUNTY', 'YEAR'])['F_DRUGOFF'].transform('sum')
ex4_num_rows['total_violent_by_race'] = violent_year
ex4_num_rows['total_drug_by_race'] = drug_year
ex4_num_rows

YEAR   RACE               COUNTY  VIOLENT  PROPERTY  F_DRUGOFF  \
0     1980  Black       Alameda County     2594      5138       1489   
1     1980  Black         Butte County       32        33          5   
2     1980  Black     Calaveras County        0         0          0   
3     1980  Black        Colusa County        0         5          0   
4     1980  Black  Contra Costa County      720      1422        368   
...    ...    ...                  ...      ...       ...        ...   
9697  2021  White        Tulare County      524       407        195   
9698  2021  White      Tuolumne County      182        63         35   
9699  2021  White       Ventura County      574       372        308   
9700  2021  White          Yolo County      171       163         51   
9701  2021  White          Yuba County      215       123         68   

      F_SEXOFF  F_ALLOTHER  F_TOTAL  M_TOTAL  S_TOTAL  total_violent_by_race  \
0           94         541     9856    24081      555                   4504   
1            0           7       77      139        0                    306   
2            0           0        0        7        0                     52   
3            0           0        5       13        0                     28   
4           24         257     2791     5201       40                   1596   
...        ...         ...      ...      ...      ...                    ...   
9697        10         413     1549     3368       31                   2360   
9698        13         289      582     1393        2                    229   
9699        13         672     1939     7509       18                   1845   
9700         8         173      566     1051        0                    480   
9701        20         138      564     1160        0                    355   

      total_drug_by_race  
0                   3569  
1                    149  
2                     83  
3                      9  
4                   1089  
...                  ...  
9697                 635  
9698                  44  
9699                 752  
9700                  92  
9701                 112  

[9702 rows x 13 columns]